From 07fcb90d30b3e52c6c66ea4568ab7b0c0084d4c1 Mon Sep 17 00:00:00 2001
From: Eric Sauvageau <merlin@asuswrt-merlin.net>
Date: Sun, 23 Aug 2026 15:52:37 -0400
Subject: [PATCH] httpd: webui: fix Wireless Log in WISP mode; enhance client
 listing in Repeater mode

WISP mode, in addition to Repeater, uses the main BSS (wl%d_) for the
uplink connection to the other router.  Parse these as a special subunit,
and display them on the webui in a "Link" category, with regular entries
then grouped in a "Local Clients" category.

A "fake" guest SSID of "Main BSS" is used so the webui can identify
it, and handle it properly.

Also implement fronthaul handling (untested) from upstream decompile.

Finally, use eval() instead of doSystem() for the initial lease copy.

(cherry picked from commit 29ab0f149778729871c10f35344188b093c03762)
---
 .../router/httpd/sysdeps/web-broadcom-am.c    | 37 ++++++++++++++-----
 .../src/router/www/Main_WStatus_Content.asp   | 19 ++++++----
 2 files changed, 40 insertions(+), 16 deletions(-)

diff --git a/release/src/router/httpd/sysdeps/web-broadcom-am.c b/release/src/router/httpd/sysdeps/web-broadcom-am.c
index d9ef27d09e..2130f7ff18 100644
--- a/release/src/router/httpd/sysdeps/web-broadcom-am.c
+++ b/release/src/router/httpd/sysdeps/web-broadcom-am.c
@@ -370,13 +370,22 @@ ej_wl_unit_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 	size_t mtl_sz = 0;
 	char leasefile_path[128] = {0};
 #endif
+#ifdef RTCONFIG_MULTILAN_MWL
+	char fh_prefix[] = "wlXXXXXXXXXX_";
+#endif
 
 #ifdef RTCONFIG_PROXYSTA
 	if (psta_exist_except(unit))
 		return ret;
 #endif
 
-	snprintf(prefix, sizeof(prefix), "wl%d_", unit);
+#ifdef RTCONFIG_MULTILAN_MWL
+	if (get_fh_if_prefix_by_unit(unit, fh_prefix, sizeof(fh_prefix)) != 0)
+		snprintf(prefix, sizeof(prefix), "%s_", fh_prefix);
+	else
+#endif
+		snprintf(prefix, sizeof(prefix), "wl%d_", unit);
+
 	name = nvram_safe_get(strcat_r(prefix, "ifname", tmp));
 
 	wl_ioctl(name, WLC_GET_RADIO, &val, sizeof(val));
@@ -397,7 +406,8 @@ ej_wl_unit_status_array(int eid, webs_t wp, int argc, char_t **argv, int unit)
 		if (nvram_match(strcat_r(prefix, "lazywds", tmp), "1") ||
 			nvram_invmatch(strcat_r(prefix, "wds", tmp), ""))
 			ret += websWrite(wp, "\"Hybrid\"");
-		else	ret += websWrite(wp, "\"AP\"");
+		else
+			ret += websWrite(wp, "\"AP\"");
 	}
 	else if (nvram_match(strcat_r(prefix, "mode", tmp), "wds"))
 	{
@@ -486,7 +496,7 @@ sta_list:
 	   cases where a device uses a static IP rather than DHCP */
 #ifdef RTCONFIG_MULTILAN_CFG
 	/* Merge all SDN lease files into one */
-	doSystem("cp /var/lib/misc/dnsmasq.leases %s", MERGED_LEASE_FILE);
+	eval("cp", "/var/lib/misc/dnsmasq.leases", MERGED_LEASE_FILE);
 	pmtl = (MTLAN_T *)INIT_MTLAN(sizeof(MTLAN_T));
 	if (pmtl) {
 		get_mtlan(pmtl, &mtl_sz);
@@ -514,22 +524,29 @@ sta_list:
 #endif
 
 
-/*** Do all subunit client lists - subunit 1 = main interface ***/
+/*** Do all subunit client lists - subunit 0 = repeater/WISP link on main BSS, subunit 1 = main interface ***/
 
 	maxunit = wl_max_no_vifs(unit);
-	for (subunit = 1; subunit < maxunit; subunit++) {
+	for (subunit = 0; subunit < maxunit; subunit++) {
 #ifdef RTCONFIG_WIRELESSREPEATER
 		if ((nvram_get_int("sw_mode") == SW_MODE_REPEATER)
 			&& (unit == nvram_get_int("wlc_band")) && (subunit == 1))
 			break;
 #endif
-		snprintf(prefix, sizeof(prefix), "wl%d.%d_", unit, subunit);
+		if (subunit == 0)
+			snprintf(prefix, sizeof(prefix), "wl%d_", unit);
+		else
+			snprintf(prefix, sizeof(prefix), "wl%d.%d_", unit, subunit);
+
 		if (nvram_match(strcat_r(prefix, "bss_enabled", tmp), "1"))
 		{
-			snprintf(name_vif, sizeof(name_vif), "wl%d.%d", unit, subunit);
+			if (subunit == 0)	// Main BSS
+				snprintf(name_vif, sizeof(name_vif), "wl%d", unit);
+			else
+				snprintf(name_vif, sizeof(name_vif), "wl%d.%d", unit, subunit);
 
 // Not primary interface - retrieve ssid and VLAN
-			if (subunit != 1) {
+			if (subunit > 1) {
 				strlcpy(guestssid, nvram_pf_safe_get(prefix, "ssid"), sizeof(guestssid));
 #ifdef RTCONFIG_MULTILAN_CFG
 				guestvlan = get_apg_vid_by_ifname(name_vif);
@@ -712,7 +729,9 @@ sta_list:
 
 // If not a Guest Network then don't push SSID and VLAN in client list
 				if (subunit == 1)
-			                ret += websWrite(wp, "\"\",\"\"],");
+					ret += websWrite(wp, "\"\",\"\"],");
+				else if (subunit == 0)	// identify Main BSS
+					ret += websWrite(wp, "\"Main BSS\",\"\"],");
 				else {
 // SSID (for Guest Networks identification)
 					ret += websWrite(wp, "\"%s\",", guestssid);
diff --git a/release/src/router/www/Main_WStatus_Content.asp b/release/src/router/www/Main_WStatus_Content.asp
index 5c8e2f76b7..0843daa6a8 100644
--- a/release/src/router/www/Main_WStatus_Content.asp
+++ b/release/src/router/www/Main_WStatus_Content.asp
@@ -186,6 +186,7 @@ function generate_clients(clientsarray) {
 	var nmapentry;
 	var guestheader = "";
 	var popupHandler;
+	var mainBSS = 0;
 
 	if (isSupport("UI4")){
 		var guestBackgroundColor = "background-color: #EEEEEE;";
@@ -213,15 +214,19 @@ function generate_clients(clientsarray) {
 			// Display a new Guest header?
 			if (client[12] != "" && client[12] != guestheader) {
 				guestheader = client[12];
-				if (sw_mode == "2") {
-					code += `<tr><th colspan="6" style="height:20px; ${guestBackgroundColor}"><span class="hint-color" style="font-weight:bolder;">Local Clients: </span> ${guestheader}`;
+				if (client[12] == "Main BSS") {
+					code += `<tr><th colspan="6" style="height:20px; ${guestBackgroundColor}"><span class="hint-color" style="font-weight:bolder;">Link</span>`;
+					mainBSS = 1;
 				} else {
 					code += `<tr><th colspan="6" style="height:20px; ${guestBackgroundColor}"><span class="hint-color" style="font-weight:bolder;">Guest Network: </span> ${guestheader}`;
-					if (client[13] != "") {
-						code += '<span style="float:right;"><span class="hint-color" style="font-weight:bolder;">VLAN: </span> ' + client[13] + '</span>';
-					}
+				}
+				if (client[13] != "") {
+					code += '<span style="float:right;"><span class="hint-color" style="font-weight:bolder;">VLAN: </span> ' + client[13] + '</span>';
 				}
 				code += '</th></tr>';
+			} else if (client[12] == "" && mainBSS == 1) {
+				code += `<tr><th colspan="6" style="height:20px; ${guestBackgroundColor}"><span class="hint-color" style="font-weight:bolder;">Local Clients</span>`;
+				mainBSS = 0;
 			}
 			code += '<tr>';
 
@@ -322,8 +327,8 @@ function generate_header(dataarray, title, dfs_statusarray) {
 		code += '<td><span class="hint-color">SNR: </span>' + dataarray[2] +' dB</td>';
 	if (dataarray[3] != 0)
 		code += '<td><span class="hint-color">Noise: </span>' + dataarray[3] + ' dBm</td>';
-        if (dataarray[6] != -1)
-                code += '<td><span class="hint-color">Utilization: </span>'+ dataarray[6] + '%</td>';
+	if (dataarray[6] != -1)
+		code += '<td><span class="hint-color">Utilization: </span>'+ dataarray[6] + '%</td>';
 	code += '</tr>';
 
 	if (dfs_statusarray.length > 1) {
-- 
2.25.1

